stan/math/prim/arr/functor/mpi_cluster.hpp

```diff
@@ -0,0 +1,251 @@
+#ifdef STAN_MPI
+
+#ifndef STAN_MATH_PRIM_ARR_FUNCTOR_MPI_CLUSTER_HPP
+#define STAN_MATH_PRIM_ARR_FUNCTOR_MPI_CLUSTER_HPP
+
+#include <stan/math/prim/arr/functor/mpi_command.hpp>
+
+#include <boost/mpi/allocator.hpp>
+#include <boost/mpi/collectives.hpp>
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/datatype.hpp>
+#include <boost/mpi/environment.hpp>
+#include <boost/mpi/nonblocking.hpp>
+#include <boost/mpi/operations.hpp>
+
+#include <boost/serialization/access.hpp>
+#include <boost/serialization/base_object.hpp>
+#include <boost/serialization/export.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+
+#include <mutex>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Exception used to stop workers nodes from further listening to
+ * commands send from the root.
+ */
+class mpi_stop_listen : public std::exception {
+  virtual const char* what() const throw() {
+    return "Stopping MPI listening mode.";
+  }
+};
+
+/**
+ * Exception thrown whenever the MPI resource is busy
+ */
+class mpi_is_in_use : public std::exception {
+  virtual const char* what() const throw() { return "MPI resource is in use."; }
+};
+
+/**
+ * MPI command used to stop childs nodes from listening for
+ * further commands.
+ */
+struct mpi_stop_worker : public mpi_command {
+  friend class boost::serialization::access;
+  template <class Archive>
+  void serialize(Archive& ar, const unsigned int version) {
+    ar& BOOST_SERIALIZATION_BASE_OBJECT_NVP(mpi_command);
+  }
+  void run() const {
+    boost::mpi::communicator world;
+    // TODO(wds15): Remove debugging print
+    std::cout << "Terminating worker " << world.rank() << std::endl;
+    throw mpi_stop_listen();
+  }
+};
+
+/**
+ * Maps jobs of given chunk size to workers and returning a vector
+ * of counts. The returned vector is indexed by the rank of each
+ * worker and has size equal to the # of workers. Each count per
+ * worker is the product of the number of assigned jobs times the
+ * chunk size. The jobs are deterministically assigned to
+ * workers. This is used for static scheduling of jobs internally.
+ *
+ * So with num_workers workers, then the counts for worker with
+ * given rank is
+ *
+ * chunks[rank] = floor(num_jobs/num_workers) * chunk_size
+ *
+ * The remainder jobs num_jobs % num_workers are assigned to rank
+ * >=1 workers such that the root (rank = 0) has a little less
+ * assigned chunks.
+ *
+ * @param num_jobs Total number of jobs to dispatch
+ * @param chunk_size Chunk size per job
+ * @return vector indexed by rank with the total number of
+ * elements mapped to a given worker
+ */
+inline std::vector<int> mpi_map_chunks(std::size_t num_jobs,
+                                       std::size_t chunk_size = 1) {
+  boost::mpi::communicator world;
+  const std::size_t world_size = world.size();
+
+  std::vector<int> chunks(world_size, num_jobs / world_size);
+
+  for (std::size_t r = 0; r != num_jobs % world_size; r++)
```

Comment:
[optional]

unless the pre-incremented value is needed, use `++r` instead of `r++`

Reply:
OK